packages/react-async/src/context/assets.ts

```diff
@@ -1,23 +1,71 @@
 import * as React from 'react';
 import {EffectKind} from '@shopify/react-effect';
+import {AssetTiming} from '../types';
+
+export interface AssetSelector {
+  id: string;
+  styles: boolean;
+  scripts: boolean;
+}
+
+interface AssetOptions {
+  styles: AssetTiming;
+  scripts: AssetTiming;
+}
 
 export const EFFECT_ID = Symbol('react-async');
 
 export class AsyncAssetManager {
-  readonly used = new Set<string>();
-
   readonly effect: EffectKind = {
     id: EFFECT_ID,
     betweenEachPass: () => {
-      this.used.clear();
+      this.assets.clear();
     },
   };
 
-  markAsUsed(id: string) {
-    this.used.add(id);
+  private assets = new Map<string, AssetOptions>();
+
+  used(timing: AssetTiming | AssetTiming[] = AssetTiming.Immediate) {
+    const timingArray = Array.isArray(timing) ? timing : [timing];
+
+    const assets: AssetSelector[] = [];
+
+    for (const [asset, {scripts, styles}] of this.assets) {
```

Comment:
spent a while parsing this for why it can't be a map/filter pair, not a big deal but might be more consistent with our other code /shrug

Reply:
Personally, I'd like to see this style become a bit more common, it's totally legit and works regardless of what type `assets` is, assuming it is utterable.